### PR TITLE
Ed/display nearest buildings

### PR DESCRIPTION
**Pull request by esmason**
This PR adds the following functionality:
- remove autopublish and use explicit pub/sub to access DB data from client
- simplify front end by refactoring component state to App.state (i.e. take state out of presentational components)
- added marker component to facilitate the previous point (and also to let react deal with re-render logic)
- other small changes such as stylistic fixes (whitespace etc.)

I would like to add [redux](https://github.com/reactjs/redux) to our app because I think it will make things easier as the front end becomes more complex. I can either do it on this branch on make a new branch and implement it after merging this one. I imagine there will be a moderate amount of changes required to do so. (maybe as many as there are currently in this PR).

### CODE REVIEW

File: imports/lib/google-map.js [hunk 1]
```diff
@@ -2,56 +2,96 @@ import React, { PropTypes } from 'react';
 import { Random } from 'meteor/random';
 import { createContainer } from 'meteor/react-meteor-data';
 import { Meteor } from 'meteor/meteor'
+import Marker from '../ui/marker'
```
Comment (JordanRoth) [inline imports/lib/google-map.js]: Should we put this in curly braces for consistency?

File: imports/lib/google-map.js [hunk 2]
```diff
@@ -2,56 +2,96 @@ import React, { PropTypes } from 'react';
...
+            });
+            this.props.onReady(this.name);
+            clearInterval(this.timer);
+        }
+    }
+
+    componentWillUnmount() {
+        if (GoogleMaps.maps[this.name]) {
+            google.maps.event.clearInstanceListeners(
+                GoogleMaps.maps[this.name].instance,
```
Comment (JordanRoth) [inline imports/lib/google-map.js:35]: Weird syntax here. Shouldn't the comma either not exist, or there be a second argument?

File: imports/lib/google-map.js [hunk 3]
```diff
@@ -2,56 +2,96 @@ import React, { PropTypes } from 'react';
...
+            <div className="map-container" ref={c => (this.container = c)}>
+                {this.props.children}
+                <div>{this.renderUserLocation()}</div>
+                <div>{this.renderBuildingMarkers()}</div>
+            </div>
+        );
+    }
+
+    renderUserLocation(){
+        if (this.props.userLocation!=null) {
```
Comment (JordanRoth) [inline imports/lib/google-map.js]: Need space around the comparison operator. If you're checking if something is null in Javascript, you can (and probably should) just write: `if (!var)` as it's cleaner.

File: imports/lib/google-map.js [hunk 4]
```diff
@@ -2,56 +2,96 @@ import React, { PropTypes } from 'react';
...
+                           map = {GoogleMaps.maps[this.name].instance}>
+            </Marker>)
+        }
+    }
+
+    renderBuildingMarkers() {
+        return this.props.buildings.map((building) =>(
+            <Marker key = {Random.id()}
+                    marker = {this.initMarker({
+                            lat:building.latitude,
```
Comment (JordanRoth) [inline imports/lib/google-map.js]: Missing space after :

File: imports/lib/google-map.js [hunk 5]
```diff
@@ -2,56 +2,96 @@ import React, { PropTypes } from 'react';
...
+                            lng: building.longitude,
+                        }
+                    )}
+                    map = {GoogleMaps.maps[this.name].instance}></Marker>
+        ));
+    }
+
+    initMarker(position){
```
Comment (JordanRoth) [inline imports/lib/google-map.js]: Missing space before {.

File: imports/ui/App.jsx [hunk 6]
```diff
@@ -3,24 +3,43 @@ import { createContainer } from 'meteor/react-meteor-data';
 import SearchBar from './search-bar.jsx';
 import MyMap from './my-map.jsx';
 
+availableBuildings = new Mongo.Collection('available-buildings');
+
+
```
Comment (JordanRoth) [inline imports/ui/App.jsx]: Extra line

File: imports/ui/App.jsx [hunk 7]
```diff
@@ -3,24 +3,43 @@ import { createContainer } from 'meteor/react-meteor-data';
...
-            userLocation: null
+            userLocation: null,
+            buildings : [],
+            subscription: null,
         };
         this.handleMapClick = this.handleMapClick.bind(this);
     }
 
     handleMapClick(event) {
+        if (this.state.subscription!=null) {
```
Comment (JordanRoth) [inline imports/ui/App.jsx]: Same as above

File: imports/ui/App.jsx [hunk 8]
```diff
@@ -3,24 +3,43 @@ import { createContainer } from 'meteor/react-meteor-data';
...
         this.setState({
-            userLocation: event.latLng
+            userLocation: event.latLng,
+            subscription: Meteor.subscribe(
+                "buildings",
+                event.latLng.lat(),
+                event.latLng.lng(),
+                () => this.setState({
+                    buildings:(typeof availableBuildings.find({}).fetch()[0] !== "undefined")?
+                        availableBuildings.find({}).fetch()[0].buildings : [],
```
Comment (JordanRoth) [inline imports/ui/App.jsx]: Spaces around ternary operators and after :

File: imports/ui/App.jsx [hunk 9]
```diff
@@ -3,24 +3,43 @@ import { createContainer } from 'meteor/react-meteor-data';
...
+                }),
+            ),
         });
     }
 
-    getUserLoc(){
-        if (this.state.userLocation){
+
```
Comment (JordanRoth) [inline imports/ui/App.jsx]: Extra line

File: imports/ui/App.jsx [hunk 10]
```diff
@@ -3,24 +3,43 @@ import { createContainer } from 'meteor/react-meteor-data';
...
+    getUserLocation(){
+        if (this.state.userLocation) {
             return "Lat: ".concat(this.state.userLocation.lat(), " Lng: ",this.state.userLocation.lng());
```
Comment (JordanRoth) [inline imports/ui/App.jsx]: Better to use the + operator or string interpolation. This is confirmed by airbnb.

File: imports/ui/App.jsx [hunk 11]
```diff
@@ -32,17 +51,18 @@ class App extends Component {
         return (
             <div className="container">
                 <header>
-                    <h1>{this.getUserLoc()}</h1>
+                    <h1>{this.getUserLocation()}</h1>
                     <SearchBar placeholder = "search for data"/>
                 </header>
-                <MyMap onMapClick = {this.handleMapClick}/>
+                <MyMap onMapClick = {this.handleMapClick}
+                       userLocation={this.state.userLocation}
```
Comment (JordanRoth) [inline imports/ui/App.jsx]: Space around = operator.

File: imports/ui/marker.jsx [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import React, { Component, PropTypes } from 'react';
+
+export default class Marker extends React.Component {
+    
+    componentDidMount() {
+        this.props.marker.setMap(this.props.map);
+    }
+
+    componentWillUnmount(){
```
Comment (JordanRoth) [inline imports/ui/marker.jsx]: Space before {

File: imports/ui/marker.jsx [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+        this.props.marker.setMap(null);
+    }
+
+    render(){
```
Comment (JordanRoth) [inline imports/ui/marker.jsx]: Space before {

File: imports/ui/my-map.jsx [hunk 14]
```diff
@@ -2,55 +2,47 @@ import React from 'react';
...
     constructor(props){
         super(props);
-        this.clickListener = this.clickListener.bind(this);
     }
 
     render() {
         return (
-            <GoogleMap onReady={this.handleOnReady.bind(this)} mapOptions={this.handleMapOptions}>
+            <GoogleMap
+                onReady={this.handleOnReady.bind(this)}
```
Comment (JordanRoth) [inline imports/ui/my-map.jsx]: 3x space around = operator.

File: imports/ui/my-map.jsx [hunk 15]
```diff
@@ -2,55 +2,47 @@ import React from 'react';
...
             zoom: 16,
         });
     }
 
     handleOnReady(name) {
         GoogleMaps.ready(name, map => {
-            map.instance["marker"] = null;
-            google.maps.event.addListener(map.instance, "click", (event) => this.clickListener(event, map));
+            google.maps.event.addListener(map.instance, "click", (event) => this.props.onMapClick(event));
             //  getDirections(map, "Life sciences centre, UBC, Vancouver", "ICICS, UBC, Vancouver");
```
Comment (JordanRoth) [inline imports/ui/my-map.jsx]: Remove commented out code.

File: server/main.js [hunk 16]
```diff
@@ -22,3 +22,22 @@ Meteor.methods({
      */
     'getClosestAvailableBuildings': getClosestAvailableBuildings
 });
+
+if (Meteor.isServer) {
+    Meteor.publish(
+        'buildings',
+        function buildingsPublication(lat, lng) {
+            console.log("got buildings");
```
Comment (JordanRoth) [inline server/main.js]: Remove console.log

**Review by JordanRoth — CHANGES_REQUESTED** [review]
I don't know enough about React to give better feedback on the actual content, so this is mostly style feedback. Maybe try `reformat code` instead of `auto-indent lines` on IntelliJ.

File: imports/lib/google-map.js (hunk 2 above)
Comment (JordanRoth) [reply imports/lib/google-map.js:35]: This is airbnb compliant. Disregard.

File: imports/lib/google-map.js (hunk 1 above)
Comment (esmason) [reply imports/lib/google-map.js]: if you want to do this have to wrap export with `{[export]}`.

seems like this is not typically done for components https://github.com/meteor/simple-todos-react/blob/master/imports/ui/App.jsx#L8

but will add file extension

**Review by JordanRoth — APPROVED** [review]

**Pull request merged**